lib/mqtt.c

```diff
@@ -143,32 +143,147 @@ static int mqtt_getsock(struct Curl_easy *data,
+  /* protocol name */
+  packet[4] = 'M';
+  packet[5] = 'Q';
+  packet[6] = 'T';
+  packet[7] = 'T';
+  /* protocol level */
+  packet[8] = 0x04;
+  /* CONNECT flag: CleanSession */
+  packet[9] = 0x02;
+  packet[10] = 0x00;
+  /* keep-alive 0 = disabled */
+  packet[11] = 0x3c;
+  packet[12] = 0x00;
+  /* payload1 length */
+  packet[13] = 0x00;
+  packet[client_id_offset - 1] = MQTT_CLIENTID_LEN;
+}
+
 static CURLcode mqtt_connect(struct Curl_easy *data)
 {
   CURLcode result = CURLE_OK;
   const size_t client_id_offset = 14;
-  const size_t packetlen = client_id_offset + MQTT_CLIENTID_LEN;
+  size_t packetlen = client_id_offset + MQTT_CLIENTID_LEN;
   char client_id[MQTT_CLIENTID_LEN + 1] = "curl";
   const size_t clen = strlen("curl");
-  char packet[32] = {
-    MQTT_MSG_CONNECT,  /* packet type */
-    0x00,              /* remaining length */
-    0x00, 0x04,        /* protocol length */
-    'M','Q','T','T',   /* protocol name */
-    0x04,              /* protocol level */
-    0x02,              /* CONNECT flag: CleanSession */
-    0x00, 0x3c,        /* keep-alive 0 = disabled */
-    0x00, 0x00         /* payload1 length */
-  };
-  packet[1] = (packetlen - 2) & 0x7f;
-  packet[client_id_offset - 1] = MQTT_CLIENTID_LEN;
+  char *packet = NULL;
+
+  /*extracting username from request*/
+  const char *username = data->state.aptr.user ?
+      data->state.aptr.user : "";
+  const size_t ulen = strlen(username);
+  /*position where starts the user payload*/
+  const size_t start_user = packetlen;
+  /*extracting password from request*/
+  const char *passwd = data->state.aptr.passwd ?
+      data->state.aptr.passwd : "";
+  const size_t plen = strlen(passwd);
+  /*position where starts the password payload*/
+  size_t start_pwd = packetlen + ulen;
+  if(ulen > 0)
+    start_pwd += 2;
+
+  packetlen = packetlen + ulen + plen;
+  /*The plus 2 are for the MSB and LSB describing the length
+   * of the string to be added on the payload. Refer to spec 1.5.2
+   * and 1.5.4 */
+  if(ulen > 0)
+    packetlen += 2;
+  if(plen > 0)
+    packetlen += 2;
+
+  /*allocating packet, given that the username and password
+   * are supplied*/
+  packet = malloc(packetlen);
+  if(!packet)
+    return CURLE_OUT_OF_MEMORY;
+  /*setting initial values for CONN pack*/
+  init_connpack(packet, packetlen);
 
   result = Curl_rand_hex(data, (unsigned char *)&client_id[clen],
                          MQTT_CLIENTID_LEN - clen + 1);
   memcpy(&packet[client_id_offset], client_id, MQTT_CLIENTID_LEN);
   infof(data, "Using client id '%s'\n", client_id);
+
+  /*if user was provided then added to the packet*/
+  if(ulen > 0)
+    add_user(username, ulen, packet, start_user);
+
+  /*if passwd was provided then added to the packet */
+  if(plen > 0)
+    add_passwd(passwd, plen, packet, start_pwd);
+
   if(!result)
     result = mqtt_send(data, packet, packetlen);
+
+  free(packet);
+  Curl_safefree(data->state.aptr.user);
+  Curl_safefree(data->state.aptr.passwd);
```

Comment:
Before freeing I would overwrite the username, password and packet with zeros using memset. Otherwise other programs might be able to read it when they acquire heap memory and don't initialize it.

Reply:
Didn't knew that either, I'm going to add that too. I'm also trying to understand how to make test cases in curl so it may take me couple of days.

Reply:
@emilengler we don't clear user name/password anywhere else so I don't think we can ask that of @geoffbeier. Also, it's not clear how that helps or even how to do it to be sure it would happen.